src/compiler/checker.ts

```diff
@@ -15848,6 +15859,19 @@ namespace ts {
                 // Eagerly resolve the constraint type which forces an error if the constraint type circularly
                 // references itself through one or more type aliases.
                 getConstraintTypeFromMappedType(type);

```

Comment:
do you perhaps understand why this isn't sort of handled within `getConstraintTypeFromMappedType`? I recall we were trying to investigate that

Reply:
So `getConstraintTypeFromMappedType` is delegating to `getConstraintOfTypeParameter` the job of getting the constraint type and additionaly sets it in the mapped type, here I have 2 choices either extend `getConstraintTypeFromMappedType` to detect the condition and eventually mutate the `constraint` of the type parameter or to extend the logic of `getConstraintOfTypeParameter` to check if the constraint is for a homomorphic mapped type. The function `getConstraintOfTypeParameter` is also directly invoked when checking index expressions like `A[k]` where `getConstraintTypeFromMappedType` isn't, the function returns the already resolved constraint if there so in theory it could still work but I felt more appropriate to have this logic in `getConstraintOfTypeParameter` as the lowest common